File: backend-dotnet/BackendDotNet/src/BackendDotNet.Library/Service/Impl/StockServiceImpl.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Transactions;
using BackendDotNet.Common.NHibernate;
using BackendDotNet.Domain;
using BackendDotNet.Dto;
using BackendDotNet.Repository;
using BackendDotNet.Service;
using CsvHelper;

namespace BackendDotNet.Service.Impl {
	public class StockServiceImpl : IStockService {
		private StockRepository stockRepository { get; set;}

		public StockServiceImpl ( StockRepository stockRepository ) {
			this.stockRepository = stockRepository;
		}

		public Object getHistoricalStockData(string symbol) {
			DateTime startDate = DateTime.Now.AddDays (-30);
			DateTime endDate = DateTime.Now;
			string strStartDate = startDate.ToString ("yyyy-MM-dd");
			string strEndDate = endDate.ToString ("yyyy-MM-dd");

			string yqlURL = "http://query.yahooapis.com/v1/public/yql";
			string dataFormat = "&format=json&env=store://datatables.org/alltableswithkeys";
			string parameters = "?q=select * from yahoo.finance.historicaldata where symbol =\"" + symbol + "\"and startDate=\"" + strStartDate + "\" and endDate=\"" + strEndDate + "\"" + dataFormat;

			HttpClient client = new HttpClient ();
			client.BaseAddress = new Uri ( yqlURL );
			client.DefaultRequestHeaders.Accept.Add(
				new MediaTypeWithQualityHeaderValue("application/json"));
			
			HttpResponseMessage response = client.GetAsync ( parameters ).Result;

			if (response.IsSuccessStatusCode) {
				var responseData = response.Content.ReadAsAsync<Object> ().Result;
				return responseData;
			} else {
				return null;
			}
		}

		public IList<StockDto> ImportStocksByCSVFile(string file) {
			var stocks = new List<StockDto> ();
			var textReader = File.OpenText ( file );  
			var parser = new CsvParser (textReader);
			var rowNumber = 0;

			using( var uow = this.stockRepository.Uow ) {
				uow.OpenSession();

				while( true ) {
					var stock = new Stock ();
					var row = parser.Read();

					if (rowNumber == 0) {
						rowNumber++;
						continue;
					} else {
						if (row == null) {
							break;
						} else {
							stock.Symbol = row [0];
							stock.Name = row [1];
							stock.LastSale = row [2];
							stock.MarketCap = row [3];
							stock.IpoYear = row [4];
							stock.Sector = row [5];
							stock.Industry = row [6];
							stock.Summary = row [7];

							this.stockRepository.Add ( stock );
							rowNumber++;
						}
					}
				}

				stocks = this.stockRepository.All ().ToList ().Select (x => new StockDto (x)).ToList ();
				uow.Commit ();
			}

			return stocks;
		}

		public List<StockDto> findStocksByWildCard(string phrase) {
			var stocks = new List<StockDto> ();

			using( var uow = this.stockRepository.Uow ) {
				uow.OpenSession ();

				if (!String.IsNullOrEmpty (phrase)) {
					StringBuilder sb = new StringBuilder ();
					sb.Append (phrase.ToUpper ()).Append ("%");
					string searchPhrase = sb.ToString ();
					stocks = this.stockRepository.FindByWildCard (searchPhrase).ToList().Select( x => new StockDto(x)).ToList();
				}

				uow.Commit ();
			}
							
			return stocks;
		}
	}
}

File: backend-java/src/main/java/com/stocks/sample/service/impl/StockServiceImpl.java
package com.stocks.sample.service.impl;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.opencsv.CSVReader;
import com.stocks.sample.domain.Stock;
import com.stocks.sample.dto.StockDto;
import com.stocks.sample.repository.StockRepository;
import com.stocks.sample.service.StockService;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.io.FileReader;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

@Service("stockService")
public class StockServiceImpl implements StockService {
    private static final Logger logger = LoggerFactory.getLogger(StockServiceImpl.class);

    @Autowired
    private StockRepository stockRepository;

    /**
     * @param file
     * @return
     * @throws Exception
     */
    public List<StockDto> importStocksByCSVFile(String file) throws Exception {
        CSVReader reader = new CSVReader(new FileReader(file));
        String[] nextLine = reader.readNext();
        List<Stock> addedStocks = new ArrayList<Stock>();

        while ((nextLine = reader.readNext()) != null) {
            Stock stock = new Stock();

            String symbol = nextLine[0];
            String name = nextLine[1];
            String lastSale = nextLine[2];
            String marketCap = nextLine[3];
            String ipoYear = nextLine[4];
            String sector = nextLine[5];
            String industry = nextLine[6];
            String summaryQuote = nextLine[7];

            List<Stock> stocks = stockRepository.findBySymbol(symbol);
            if (stocks.size() > 1) {
                throw new Exception("there should be only one record for stock symbol: " + symbol);
            } else if (stocks.size() == 1) {
                stock = stocks.get(0);
            }

            stock.setSymbol(symbol);
            stock.setName(name);
            stock.setLastSale(lastSale);
            stock.setMarketCap(marketCap);
            stock.setIpoYear(ipoYear);
            stock.setSector(sector);
            stock.setIndustry(industry);
            stock.setSummary(summaryQuote);

            addedStocks.add(stockRepository.saveAndFlush(stock));

        }
        List<StockDto> results = stockEntitiesToStockDtos(addedStocks);
        return results;
    }

    /**
     * @param phrase
     * @return
     */
    public List<StockDto> findStocksByWildCard(String phrase) {
        List<StockDto> results = new ArrayList<StockDto>();

        if (!StringUtils.isEmpty(phrase) && phrase.length() > 0) {
            StringBuilder sb = new StringBuilder();
            sb.append(phrase.toUpperCase()).append("%");
            String searchPhrase = sb.toString();
            List<Stock> stocks = stockRepository.findByWildCard(searchPhrase);
            results = stockEntitiesToStockDtos(stocks);
        }

        return results;
    }

    /**
     * @param symbol
     * @return
     */
    public String getHistoricalStockData(String symbol) throws JsonProcessingException {
        Calendar cal = Calendar.getInstance();
        cal.setTime(new Date());
        cal.add(Calendar.DATE, -30);

        Date startDate = cal.getTime();
        Date endDate = new Date();

        SimpleDateFormat formatter = new SimpleDateFormat("YYYY-MM-dd");
        String strStartDate = formatter.format(startDate);
        String strEndDate = formatter.format(endDate);

        String yqlURL = "http://query.yahooapis.com/v1/public/yql?q=";
        String dataFormat = "&format=json&env=store://datatables.org/alltableswithkeys";
        String historicalQ = yqlURL + "select * from yahoo.finance.historicaldata where symbol =\"" + symbol + "\"and startDate=\"" + strStartDate + "\" and endDate=\"" + strEndDate + "\"" + dataFormat;

        RestTemplate restTemplate = new RestTemplate();
        String result = restTemplate.getForObject(historicalQ, String.class);

        return result;
    }

    /**
     * @param stocks
     * @return
     */
    private List<StockDto> stockEntitiesToStockDtos(List<Stock> stocks) {
        List<StockDto> results = new ArrayList<StockDto>();

        for (Stock stock : stocks) {
            results.add(new StockDto(stock));
        }

        return results;
    }
}
